components/script/dom/blob.rs

```diff
@@ -271,33 +270,46 @@ impl BlobMethods<crate::DomTypeHolder> for Blob {
     }
 
     // https://w3c.github.io/FileAPI/#arraybuffer-method-algo
-    fn ArrayBuffer(&self, can_gc: CanGc) -> Rc<Promise> {
-        let global = self.global();
-        let in_realm_proof = AlreadyInRealm::assert();
-        let p = Promise::new_in_current_realm(InRealm::Already(&in_realm_proof), can_gc);
+    fn ArrayBuffer(&self, in_realm: InRealm, can_gc: CanGc) -> Rc<Promise> {
+        let cx = GlobalScope::get_cx();
+        let global = GlobalScope::from_safe_context(cx, in_realm);
+        let p = Promise::new_in_current_realm(in_realm, can_gc);
 
-        let id = self.get_blob_url_id();
+        // 1. Let stream be the result of calling get stream on this.
+        let stream = self.get_stream(can_gc);
 
-        global.read_file_async(
-            id,
-            p.clone(),
-            Box::new(|promise, bytes| {
-                match bytes {
-                    Ok(b) => {
-                        let cx = GlobalScope::get_cx();
-                        let result = run_array_buffer_data_algorithm(cx, b, CanGc::note());
-
-                        match result {
-                            Ok(FetchedData::ArrayBuffer(a)) => {
-                                promise.resolve_native(&a, CanGc::note())
-                            },
-                            Err(e) => promise.reject_error(e, CanGc::note()),
-                            _ => panic!("Unexpected result from run_array_buffer_data_algorithm"),
-                        }
-                    },
-                    Err(e) => promise.reject_error(e, CanGc::note()),
-                };
+        // 2. Let reader be the result of getting a reader from stream.
+        //    If that threw an exception, return a new promise rejected with that exception.
+        let reader = match stream.and_then(|s| s.acquire_default_reader(can_gc)) {
+            Ok(r) => r,
+            Err(e) => {
+                p.reject_error(e, can_gc);
+                return p;
+            },
+        };
+
+        // 3. Let promise be the result of reading all bytes from stream with reader.
+        let p_success = p.clone();
+        let p_failure = p.clone();
+        reader.read_all_bytes(
+            cx,
+            &global,
+            Rc::new(move |bytes| {
+                rooted!(in(*cx) let mut js_object = ptr::null_mut::<JSObject>());
```

Comment:
This is missing the prose from "Return the result of transforming promise by a fulfillment handler that returns a new [ArrayBuffer](https://webidl.spec.whatwg.org/#idl-ArrayBuffer) whose contents are its first argument."